x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sample/CircuitBreakerTests.java

```diff
@@ -0,0 +1,335 @@
+import org.elasticsearch.xpack.ql.type.EsField;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.emptySet;
+import static org.elasticsearch.action.ActionListener.wrap;
+import static org.elasticsearch.index.query.QueryBuilders.matchAllQuery;
+import static org.elasticsearch.xpack.eql.execution.assembler.SampleQueryRequest.COMPOSITE_AGG_NAME;
+import static org.elasticsearch.xpack.eql.execution.sample.SampleIterator.CB_STACK_SIZE_PRECISION;
+import static org.elasticsearch.xpack.eql.plugin.EqlPlugin.CIRCUIT_BREAKER_LIMIT;
+import static org.elasticsearch.xpack.eql.plugin.EqlPlugin.CIRCUIT_BREAKER_NAME;
+import static org.elasticsearch.xpack.eql.plugin.EqlPlugin.CIRCUIT_BREAKER_OVERHEAD;
+
+public class CircuitBreakerTests extends ESTestCase {
+
+    private static final TestCircuitBreaker CIRCUIT_BREAKER = new TestCircuitBreaker();
+
+    public void testCircuitBreakerOnStackPush() {
+        SampleIterator iterator = new SampleIterator(new QueryClient() {
+            @Override
+            public void query(QueryRequest r, ActionListener<SearchResponse> l) {}
+
+            @Override
+            public void fetchHits(Iterable<List<HitReference>> refs, ActionListener<List<List<SearchHit>>> listener) {}
+        }, mockCriteria(), randomIntBetween(10, 500), CIRCUIT_BREAKER);
+
+        CIRCUIT_BREAKER.startBreaking();
+        iterator.pushToStack(new SampleIterator.Page(CB_STACK_SIZE_PRECISION / 2));
+        iterator.pushToStack(new SampleIterator.Page(CB_STACK_SIZE_PRECISION - (CB_STACK_SIZE_PRECISION / 2) - 1));
+        expectThrows(CircuitBreakingException.class, () -> iterator.pushToStack(new SampleIterator.Page(1)));
+    }
+
+    public void testMemoryCleared() {
+        testMemoryCleared(randomBoolean());
+    }
+
+    private void testMemoryCleared(boolean fail) {
+        List<BreakerSettings> eqlBreakerSettings = Collections.singletonList(
+            new BreakerSettings(
+                CIRCUIT_BREAKER_NAME,
+                CIRCUIT_BREAKER_LIMIT,
+                CIRCUIT_BREAKER_OVERHEAD,
+                CircuitBreaker.Type.MEMORY,
+                CircuitBreaker.Durability.TRANSIENT
+            )
+        );
+        try (
+            CircuitBreakerService service = new HierarchyCircuitBreakerService(
+                Settings.EMPTY,
+                eqlBreakerSettings,
+                new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS)
+            );
+            ESMockClient esClient = new ESMockClient(service.getBreaker(CIRCUIT_BREAKER_NAME));
+        ) {
+            CircuitBreaker eqlCircuitBreaker = service.getBreaker(CIRCUIT_BREAKER_NAME);
+            EqlConfiguration eqlConfiguration = new EqlConfiguration(
```

Comment:
Also here, wondering if it'd be worth adding a EqlTestUtils method to configure a configuration, since most of this is contained in the `TEST_CFG`.

Reply:
I think that's what `EqlTestUtils#randomConfiguration()` already does, actually. Double-checking to see if there are significant differences, but I guess we can just use that.